docs/proposals/petset.md

```diff
@@ -0,0 +1,533 @@
+// Downside: possibly too flexible, can conflict with existing env resolution semantics, bypasses downward API
+{
+    "spec": {
+        "template": {
+            "metadata": {"name": "pod-name-$(IDENTITY)"},
+            "spec": {
+                "containers": [{
+                    "env": [{
+                        "name": "IDENTITY",
+                        "value": "$(IDENTITY)"
+                    }]
+                }]
+            }
+        }
+    }
+}
+
+// Proposal 1e: specify merge-patch semantics with some variable replacement
+// Downside: merge patch is inflexible to structural template edits
+{
+    "spec": {
+        "template": {
+            "metadata": {},
+            "spec": {
+                "containers": [{
+                    "env": [{
+                        "name": "IDENTITY",
+                    }]
+                }]
+            }
+        },
+        "templateTransform": [
+            {"op": "add", "path": "/metadata/name", "value": "pod-name-$(IDENTITY)"},
+            {"op": "add", "path": "/spec/containers[0]/env[0]/value", "value": "$(IDENTITY)"}
+        ]
+    }
+}
+```
+
+Proposal 1e has the most overlap with other concepts already existing in the API, and makes it easy
+for clients to identify what will be transformed per template. Outside-in transformations also are
+most amenable to automatic generation.
+
+Because that pod needs access to stable storage, the PetSet may specify a template for one or more
+**supporting resources** that can be used for each distinct pod. In the future other resources may be
+added that must also be templated - for instance, secrets (unique secret per member), config data (unique
+config per member), and in the future, arbitrary extension resources.
+
+The set of parameterization allowed to each pod and supporting resources is intentionally limited to those
+that accomplish the goals above. It should be possible to:
+
+* Identify the supporting resources by name in the pod template (pvc name)
+* Give the supporting resource a predictable name based on identity
+* Set environment variables that correspond to the identity of the pod for use by the software
+* Provide the current desired set size (`spec.replicas`) to each pod as an integer
+
+```
+type PetSetSpec struct {
+    ...
+    // The template to use for the pod in the set
+    Template *PodTemplateSpec
+    SupportTemplates []SupportTemplateSpec
+}
+
+type SupportTemplateSpec struct {
+    // only one of these fields may be set, like Volumes
+    PersistentVolumeClaim *PersistentVolumeClaimTemplateSpec
+    Secret *SecretTemplateSpec
+    ConfigData *ConfigDataTemplateSpec
+    // anything not supported by the above
+    Object runtime.Object
+}
+
+// Proposal 2a: use the name of the support resource as identity implicitly
+// Downside: Requires the controller to understand the places to be parameterized in the template - prevents
+//   custom resources from being supported.
```

Comment:
I think config data is useful, but it's just a tool.  Someone who
wants to use a config data in their template doesn't need this
mechanism - they can use the existing mechanisms for that.  This would
be specifically for delivering "controller context"  into the pod
template, where context is only possible to retrieve at the time of
creation (for now, at least).  All other dynamic context should have a
separate mechanism.

At least, that made sense to me as I was justifying a template
mechanism to myself.